blocks/fileio/include/gnuradio-4.0/fileio/BasicFileIo.hpp

```diff
@@ -0,0 +1,307 @@
+        _mode = magic_enum::enum_cast<FileMode>(mode, magic_enum::case_insensitive).value_or(_mode);
+        if (lifecycle::isActive(this->state())) {
+            closeFile();
+            openNextFile();
+        }
+    }
+
+    void start() {
+        _totalBytesWritten = 0UZ;
+        openNextFile();
+    }
+
+    void stop() { closeFile(); }
+
+    [[nodiscard]] constexpr work::Status processBulk(ConsumableSpan auto& dataIn) {
+        std::size_t nBytesMax = dataIn.size() * sizeof(T);
+        if (max_bytes_per_file.value != 0U) {
+            nBytesMax = std::min(nBytesMax, static_cast<std::size_t>(max_bytes_per_file.value - _totalBytesWrittenFile));
+        }
+        _file.write(reinterpret_cast<const char*>(dataIn.data()), static_cast<std::streamsize>(nBytesMax));
+        if (!dataIn.consume(nBytesMax / sizeof(T))) {
+            throw gr::exception("could not consume input samples");
+        }
+
+        if (!_file) {
+            throw gr::exception(fmt::format("failed to write to file '{}'.", _actualFileName));
+        }
+
+        _totalBytesWritten += nBytesMax;
+        _totalBytesWrittenFile += nBytesMax;
+        if (max_bytes_per_file.value != 0U && _totalBytesWrittenFile >= max_bytes_per_file.value) {
+            closeFile();
+            openNextFile();
+        }
+
+        return work::Status::OK;
+    }
+
+private:
+    void closeFile() {
+        if (_file.is_open()) {
+            _file.flush();
+            _file.close();
+        }
+    }
+    void openNextFile() {
+        closeFile();
+        _totalBytesWrittenFile = 0UZ;
+
+        detail::ensureDirectoryExists(file_name.value);
+
+        std::filesystem::path filePath(file_name.value);
+        if (!std::filesystem::exists(filePath.parent_path())) {
+            throw gr::exception(fmt::format("path/file '{}' does not exist.", file_name.value));
+        }
+
+        // Open file handle based on mode
+        switch (_mode) {
+        case FileMode::override: {
+            _actualFileName = file_name.value;
+            _file.open(_actualFileName, std::ios::binary | std::ios::trunc);
+        } break;
+        case FileMode::append: {
+            _actualFileName = file_name.value;
+            _file.open(_actualFileName, std::ios::binary | std::ios::app);
+        } break;
+        case FileMode::multi: {
+            // _fileCounter ensures that the filenames are unique and still sortable by date-time, with an additional counter to handle rapid successive file creation.
+            _actualFileName = filePath.parent_path() / (detail::getIsoTime() + "_" + std::to_string(_fileCounter++) + "_" + filePath.filename().string());
+            _file.open(_actualFileName, std::ios::binary);
+            break;
+        }
+        default: throw gr::exception("unsupported file mode.");
+        }
+
+        if (!_file) {
+            throw gr::exception(fmt::format("failed to open file '{}'.", _actualFileName));
+        }
```

Comment:
I'm not sure if I'm right, but doesn't `_file.open()` throw an exception if the file cannot be opened? That exception would currently not be caught by this function, so it seems better to open the file in a `try` block, catch the exception, and re-raise it as a `gr::exception`.

Reply:
Sounds good.